### DISCUSSION TITLE
Integrate Authelia with a reverse proxy, e.g Caddy

### DISCUSSION THREAD

**Discussion by idavollen** [authelia/authelia#7820] [Documentation Feedback] [answered]
https://caddyserver.com/docs/caddyfile/directives/forward_auth as well as the [Expand Form](https://caddyserver.com/docs/caddyfile/directives/forward_auth#expanded-form)

# Serve the authentication gateway itself
auth.example.com {
	reverse_proxy authelia:9091
}

# Serve your app
app1.example.com {
	forward_auth authelia:9091 {
		uri /api/verify?rd=https://auth.example.com
		copy_headers Remote-User Remote-Groups Remote-Name Remote-Email
	}

	reverse_proxy app1:8080
}

This directive makes a GET request to the configured upstream with the uri rewritten:

If the upstream responds with a 2xx status code, then access is granted and the header fields in copy_headers are copied to the original request, and handling continues.
Otherwise, if the upstream responds with any other status code, then the upstream's response is copied back to the client. This response should typically involve a redirect to login page of the authentication gateway.

The above mentioned documentation is a bit absurb or difficult to understand. If the upstream responds with a 2xx status code, what is the mentioned upstream here? if it is the auth.example.com, its index page showing the Sign-in Web GUI page has already returned 200 as following:

Request URL:
https://authelia.example.com/?rd=https%3A%2F%2Ftraefik.example.com%2Frest%2Fhallos&rm=GET
Request Method:
GET
Status Code:
200 OK
Remote Address:
127.0.0.1:443
Referrer Policy:
strict-origin-when-cross-origin

Can you tell me which particular URL will Caddy check the 2xx status? Which security mechanism is agreed between Caddy and Authelia for informing the concerned user is authenticated? via REMOTE-USER response header? Is it possible to bypass it or impersonate another one by a threat actor to manipulate these kind of REMOTE-XXXX headers? Any effective countermeasures?

Or if the upstream is the app1:8080, what if the app1:8080 backend doesn't well follow the security principle, implicit-deny or even has bad access control implementation? I'm looking forward to a clearly technical elaberation!

**Comment by james-d-elliott** [comment, marked as answer]
Caddy is checking the response code of `/api/verify` in this instance, `/api/verify` is given the context of the original request from the X-Forwarded-* headers Caddy sets.

The method to secure the `Remote-User` etc headers is to ensure that the service can only be accessed via Caddy so that it can ensure these headers are only set by Caddy and not third parties. If this is not possible to ensure then we'd recommend using a mechanism designed for foreign system identity federation like OpenID Connect 1.0.
